File: windows/eScratch/MarkdownPaste.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using HtmlAgilityPack;

namespace eScratch;

// Clipboard HTML is parsed as data, never loaded into a browser or executed.
public static class MarkdownPaste
{
    private sealed record Part(string Text, string Markdown);
    private static string Escape(string text) => Regex.Replace(text, @"([\\`*_{}\[\]<>#~|])", @"\$1");
    private static string Normalize(string text) => Regex.Replace(text, @"\s+", " ").Trim();

    public static string Convert(string plain, string html, bool hasRtf)
    {
        if (string.IsNullOrWhiteSpace(html)) return plain;
        if (html.Length > 2_000_000 || plain.Length > 2_000_000) return plain;
        var fragment = Regex.Match(html, @"(?s)<!--StartFragment-->(.*?)<!--EndFragment-->");
        var doc = new HtmlDocument();
        doc.LoadHtml(fragment.Success ? fragment.Groups[1].Value : html);
        var root = doc.DocumentNode.SelectSingleNode("//body") ?? doc.DocumentNode;
        var parts = new List<Part>();
        void Mark(string value) => parts.Add(new("", value));
        void Text(string value) { foreach (char c in value) parts.Add(new(c.ToString(), Escape(c.ToString()))); }
        void Visit(HtmlNode node, int depth)
        {
            if (depth > 100) return;
            var name = node.Name.ToLowerInvariant();
            if (name is "script" or "style" or "head" or "iframe" or "object" || node.NodeType == HtmlNodeType.Comment) return;
            if (node.NodeType == HtmlNodeType.Text) { Text(Regex.Replace(HtmlEntity.DeEntitize(node.InnerText), @"\s+", " ")); return; }
            if (name == "br") { Text("\n"); return; }
            var style = node.GetAttributeValue("style", "").ToLowerInvariant();
            if (Regex.IsMatch(style, @"(?:display\s*:\s*none|visibility\s*:\s*hidden)")) return;
            bool bold = name is "b" or "strong" || Regex.IsMatch(style, @"font-weight\s*:\s*(bold|[7-9]00)");
            bool italic = name is "i" or "em" || Regex.IsMatch(style, @"font-style\s*:\s*italic");
            bool strike = name is "s" or "strike" or "del" || style.Contains("line-through");
            bool block = name is "p" or "div" or "li" or "tr" or "blockquote" || Regex.IsMatch(name, "^h[1-6]$");
            if (block && parts.Count > 0) Text("\n");
            if (Regex.IsMatch(name, "^h[1-6]$")) Mark(new string('#', name[1] - '0') + " ");
            if (name == "li") Mark(node.ParentNode.Name == "ol" ? "1. " : "- ");
            if (name == "blockquote") Mark("> ");
            var marker = (bold ? "**" : "") + (italic ? "*" : "") + (strike ? "~~" : "");
            Mark(marker);
            var href = HtmlEntity.DeEntitize(node.GetAttributeValue("href", ""));
            bool link = name == "a" && Uri.TryCreate(href, UriKind.Absolute, out var uri) && uri.Scheme is "https" or "http" or "mailto";
            if (link) Mark("[");
            foreach (var child in node.ChildNodes) Visit(child, depth + 1);
            if (link) Mark("](" + href.Replace(" ", "%20").Replace("(", "%28").Replace(")", "%29") + ")");
            Mark((strike ? "~~" : "") + (italic ? "*" : "") + (bold ? "**" : ""));
            if (name is "td" or "th") Text("\t");
            if (block) Text("\n");
        }
        Visit(root, 0);
        // Normalize whitespace only for comparison; retain the Markdown layout.
        var final = new StringBuilder(); var map = new List<int>();
        for (int i = 0; i < parts.Count; i++)
            foreach (char c in parts[i].Text)
            {
                if (char.IsWhiteSpace(c)) { if (final.Length == 0 || final[^1] == ' ') continue; final.Append(' '); }
                else final.Append(c);
                map.Add(i);
            }
        if (final.Length > 0 && final[^1] == ' ') { final.Length--; map.RemoveAt(map.Count - 1); }
        var inserts = hasRtf ? RecoverDeletions(Normalize(plain), final.ToString()) : new Dictionary<int, string>();
        var before = new Dictionary<int, string>();
        foreach (var pair in inserts) before[pair.Key < map.Count ? map[pair.Key] : map.Count > 0 ? map[^1] + 1 : parts.Count] = pair.Value;
        var result = new StringBuilder();
        for (int i = 0; i <= parts.Count; i++)
        {
            if (before.TryGetValue(i, out var deleted)) result.Append(Regex.Match(deleted, @"^\s*").Value + "~~" + Escape(deleted.Trim()) + "~~" + Regex.Match(deleted, @"\s*$").Value);
            if (i < parts.Count) result.Append(parts[i].Markdown);
        }
        return Regex.Replace(result.ToString(), @"\n[ \t]*\n(?:[ \t]*\n)+", "\n\n").Trim();
    }

    public static Dictionary<int, string> RecoverDeletions(string original, string final)
    {
        var result = new Dictionary<int, string>();
        if (final.Length == 0 || original.Length <= final.Length) return result;
        // Only accept a unique deletion-only alignment. A forward/backward scan
        // avoids quadratic diff costs and rejects ambiguous repeated text.
        var positions = new int[final.Length]; int cursor = 0;
        for (int i = 0; i < final.Length; i++)
        {
            cursor = original.IndexOf(final[i], cursor);
            if (cursor < 0) return result;
            positions[i] = cursor++;
        }
        cursor = original.Length - 1;
        for (int i = final.Length - 1; i >= 0; i--)
        {
            cursor = original.LastIndexOf(final[i], cursor);
            if (cursor != positions[i]) return result;
            cursor--;
        }
        cursor = 0;
        for (int i = 0; i <= final.Length; i++)
        {
            int end = i == final.Length ? original.Length : positions[i];
            if (end > cursor)
            {
                var deleted = original.Substring(cursor, end - cursor);
                if (!string.IsNullOrWhiteSpace(deleted)) result[i] = deleted;
            }
            cursor = end + 1;
        }
        return result;
    }
}

File: src/markdownPaste.ts
function recoverWords(original: string, final: string): Map<number, string> {
  const result = new Map<number, string>()
  const a = [...original.matchAll(/\S+/g)], b = [...final.matchAll(/\S+/g)]
  if (!b.length) return result
  const positions: number[] = []
  let cursor = 0
  for (const word of b) {
    while (cursor < a.length && a[cursor][0] !== word[0]) cursor++
    if (cursor === a.length) return result
    positions.push(cursor++)
  }
  cursor = a.length - 1
  for (let i = b.length - 1; i >= 0; i--) {
    while (cursor >= 0 && a[cursor][0] !== b[i][0]) cursor--
    if (cursor !== positions[i]) return result
    cursor--
  }
  let previous = -1
  b.forEach((word, i) => {
    if (positions[i] > previous + 1) result.set(word.index!, original.slice(a[previous + 1].index!, a[positions[i]].index!))
    previous = positions[i]
  })
  if (previous < a.length - 1) result.set(final.length, original.slice(a[previous].index! + a[previous][0].length))
  return result
}

type Part = { text: string; markdown: string }
const escape = (text: string) => text.replace(/([\\`*_{}\[\]<>#~|])/g, '\\$1')
const normalize = (text: string) => text.replace(/\s+/g, ' ').trim()

export function recoverDeletions(original: string, final: string): Map<number, string> {
  const result = new Map<number, string>()
  if (!final || original.length <= final.length) return result
  const positions: number[] = []
  let cursor = 0
  for (let i = 0; i < final.length; i++) {
    cursor = original.indexOf(final[i], cursor)
    if (cursor < 0) return result
    positions.push(cursor++)
  }
  cursor = original.length - 1
  for (let i = final.length - 1; i >= 0; i--) {
    cursor = original.lastIndexOf(final[i], cursor)
    if (cursor !== positions[i]) return recoverWords(original, final)
    cursor--
  }
  cursor = 0
  for (let i = 0; i <= final.length; i++) {
    const end = i === final.length ? original.length : positions[i]
    if (end > cursor) {
      const deleted = original.slice(cursor, end)
      if (deleted.trim()) result.set(i, deleted)
    }
    cursor = end + 1
  }
  return result
}

export function pasteAsMarkdown(plain: string, html: string, hasRtf: boolean): string {
  if (!html.trim() || html.length > 2_000_000 || plain.length > 2_000_000) return plain
  const fragment = html.match(/<!--StartFragment-->([\s\S]*?)<!--EndFragment-->/)
  // Template content is inert and never attached to the live document.
  const template = document.createElement('template')
  template.innerHTML = fragment ? fragment[1] : html
  const parts: Part[] = []
  const mark = (markdown: string) => { parts.push({ text: '', markdown }) }
  const text = (value: string) => { for (let i = 0; i < value.length; i++) parts.push({ text: value[i], markdown: escape(value[i]) }) }
  const visit = (node: Node, depth: number) => {
    if (depth > 100) return
    if (node.nodeType === Node.COMMENT_NODE) return
    if (node.nodeType === Node.TEXT_NODE) { text((node.textContent || '').replace(/\s+/g, ' ')); return }
    if (!(node instanceof Element)) { node.childNodes.forEach(n => visit(n, depth + 1)); return }
    const name = node.tagName.toLowerCase()
    if (['script', 'style', 'head', 'iframe', 'object'].includes(name)) return
    if (name === 'br') { text('\n'); return }
    const style = (node.getAttribute('style') || '').toLowerCase()
    if (/(?:display\s*:\s*none|visibility\s*:\s*hidden)/.test(style)) return
    const bold = ['b', 'strong'].includes(name) || /font-weight\s*:\s*(bold|[7-9]00)/.test(style)
    const italic = ['i', 'em'].includes(name) || /font-style\s*:\s*italic/.test(style)
    const strike = ['s', 'strike', 'del'].includes(name) || style.includes('line-through')
    const block = ['p', 'div', 'li', 'tr', 'blockquote'].includes(name) || /^h[1-6]$/.test(name)
    if (block && parts.length) text('\n')
    if (/^h[1-6]$/.test(name)) mark('#'.repeat(Number(name[1])) + ' ')
    if (name === 'li') mark(node.parentElement?.tagName === 'OL' ? '1. ' : '- ')
    if (name === 'blockquote') mark('> ')
    mark((bold ? '**' : '') + (italic ? '*' : '') + (strike ? '~~' : ''))
    const href = node.getAttribute('href') || ''
    const link = name === 'a' && /^(https?:\/\/|mailto:)/i.test(href)
    if (link) mark('[')
    node.childNodes.forEach(n => visit(n, depth + 1))
    if (link) mark('](' + href.replace(/ /g, '%20').replace(/\(/g, '%28').replace(/\)/g, '%29') + ')')
    mark((strike ? '~~' : '') + (italic ? '*' : '') + (bold ? '**' : ''))
    if (['td', 'th'].includes(name)) text('\t')
    if (block) text('\n')
  }
  visit(template.content, 0)
  let final = ''
  const map: number[] = []
  parts.forEach((part, i) => {
    for (let j = 0; j < part.text.length; j++) {
      const c = part.text[j]
      if (/\s/.test(c)) { if (!final || final.endsWith(' ')) continue; final += ' ' }
      else final += c
      map.push(i)
    }
  })
  if (final.endsWith(' ')) { final = final.slice(0, -1); map.pop() }
  const before = new Map<number, string>()
  const comparisonPlain = template.content.querySelector('li') ? plain.replace(/^[ \t]*(?:\d+[.)]|[•·])[^\S\r\n]+/gm, '') : plain
  if (hasRtf) recoverDeletions(normalize(comparisonPlain), final).forEach((value, key) => before.set(key < map.length ? map[key] : map.length ? map[map.length - 1] + 1 : parts.length, value))
  let output = ''
  for (let i = 0; i <= parts.length; i++) {
    if (before.has(i)) { const deleted = before.get(i)!; output += (deleted.match(/^\s*/)?.[0] || '') + '~~' + escape(deleted.trim()) + '~~' + (deleted.match(/\s*$/)?.[0] || '') }
    if (i < parts.length) output += parts[i].markdown
  }
  return output.replace(/\n[ \t]*\n(?:[ \t]*\n)+/g, '\n\n').trim()
}
